katharsis-core/src/main/java/io/katharsis/resource/registry/responseRepository/ResponseRepository.java

```diff
@@ -33,66 +45,218 @@
+			List<RepositoryFilter> filters = moduleRegistry.getRepositoryFilters();
+			if (filterIndex == filters.size()) {
+				return doGetLinksInformation(repository, resources, context.getRequest());
+			}
+			else {
+				RepositoryFilter filter = filters.get(filterIndex);
+				filterIndex++;
+				return filter.filterLinks(context, resources, this);
+			}
+		}
+	}
+    
+    class RepositoryResultFilterChainImpl<T> implements RepositoryResultFilterChain<T> {
+
+		protected int filterIndex = 0;
+		
+		private Iterable<T> result;
+		
+		public RepositoryResultFilterChainImpl(Iterable<T> result){
+			this.result = result;
+		}
+
+		@Override
+		public Iterable<T> doFilter(RepositoryFilterContext context) { // NOSONAR
+			List<RepositoryFilter> filters = moduleRegistry.getRepositoryFilters();
+			if (filterIndex == filters.size()) {
+				return result;
+			}
+			else {
+				RepositoryFilter filter = filters.get(filterIndex);
+				filterIndex++;
+				return filter.filterResult(context, this);
+			}
+		}
+	}
+    
+    protected abstract class RepositoryRequestFilterChainImpl implements RepositoryRequestFilterChain {
+
+		protected int filterIndex = 0;
+
+		@Override
+		public JsonApiResponse doFilter(RepositoryFilterContext context) {
+			List<RepositoryFilter> filters = moduleRegistry.getRepositoryFilters();
+			if (filterIndex == filters.size()) {
+				return invoke(context);
+			}
+			else {
+				RepositoryFilter filter = filters.get(filterIndex);
+				filterIndex++;
+				return filter.filterRequest(context, this);
+			}
+		}
+
+		protected abstract JsonApiResponse invoke(RepositoryFilterContext context);
+	}
+    
+    protected abstract class RepositoryBulkRequestFilterChainImpl<K> implements RepositoryBulkRequestFilterChain<K> {
+
+		protected int filterIndex = 0;
+
+		@Override
+		public Map<K, JsonApiResponse> doFilter(RepositoryFilterContext context) {
+			List<RepositoryFilter> filters = moduleRegistry.getRepositoryFilters();
+			if (filterIndex == filters.size()) {
+				return invoke(context);
+			}
+			else {
+				RepositoryFilter filter = filters.get(filterIndex);
+				filterIndex++;
+				return filter.filterBulkRequest(context, this);
+			}
+		}
+
+		protected abstract Map<K, JsonApiResponse> invoke(RepositoryFilterContext context);
+	}
+
+    protected RepositoryFilterContext newRepositoryFilterContext(final RepositoryRequestSpec requestSpec) {
+		return new RepositoryFilterContext(){
+
+			@Override
+			public RepositoryRequestSpec getRequest() {
+				return requestSpec;
+			}
+		};
+	}
+   
     @SuppressWarnings({ "unchecked", "rawtypes" })
-    private LinksInformation getLinksInformation(Object repository, Iterable<?> resources, RequestSpec requestSpec) {
-    	QueryAdapter queryAdapter = requestSpec.getQueryAdapter();
+    private LinksInformation doGetLinksInformation(Object repository, Iterable<?> resources, RepositoryRequestSpec requestSpec) {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Move this method into "RepositoryLinksFilterChainImpl". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AS3398)